### ISSUE TITLE
ODP compiler: Possible to specify "signer"?

### ISSUE THREAD

**Issue by eivindr** [OpenNTF/org.openntf.nsfodp#180]
Hi, is it possible to specify the "user id" used as signer for the generated NSF?

It seems like the username/password in maven settings.xml (servers:server) is not used as the signer. The signer always ends up as the "server id" here.

Is this the expected behavior or am I doing something wrong?

**Comment by jesse-gallagher** [comment]
You’re not doing anything wrong: it’s a limitation in how Domino does signing. When you sign a note, it’s always with the active Notes ID, which will be the server when doing remote compilation.

There’s a hypothetical workaround when you have IDs in a Vault, but it’s tricky and not a supported API, so I’ve been wary about adding it. I’ll keep it in mind, though, and maybe lobby HCL for a documented API to do that without hacks.

**Comment by eivindr** [comment]
I agree, some official way to sign NSF from api/command line using a supplied user id would be ideal.

We're now using a domino docker image (with nsfodp) when building our NSFs. So we can configure/"hack" the Domino installation as needed. I tried to specify the server.id in the Domino setup but it seems Domino will create a new server.id regardless.

Do you have any input on how to best solve this "now"? Not necessarily inside the nsfodp tool. Do you have any inside info on how people are solving CI/CD type workflows when Domino is involved?

**Comment by jesse-gallagher** [comment]
In my cases, the deployment step is still manual. Locally, I develop with a Liberty server that I cobbled together to run the app, so I skip that part, and then the production deployment involves taking a distribution of the OSGi plugins + NSFs and deploying them in an NSF Update Site and with Administrator, respectively. Since that's being done manually, there's a last step of signing the DB that fits in.

There _is_ some mostly-prototype code in this tooling for deployment, and I'd love to flesh that out some time for a true CD setup where builds go out to Domino. The things standing in my way at the moment are:

- Signing. This could be done with the "sign with server ID" `AdministrationProcess` call, but it'd be good to also be able to do that with an admin/developer ID from the vault.
- Upgrading the design properly. As far as I know, there's no API call equivalent to the "Replace Design" command in the Notes client, which is the behavior I'd want. I can issue a console command to refresh the design if you use a template name and I copy the template out, but that'd be problem-prone.
- Deploying OSGi bundles properly. This is largely possible - it's what the ODP Tooling does on the server at runtime, in fact - but I've run into trouble where the server won't properly unload bundles and will eventually need a reboot. Additionally, I don't know how this would affect running XPages/Equinox apps.

**Comment by eivindr** [comment]
I've discussed this some more with one of our admins and we're going to re-configure our build environment so the docker "nsf builder" container become an additional server. This way we can specify the server.id file and use the "signer id" we want.

> Upgrading the design
In our deployment script we're using the "(n)convert" command to replace design. Probably misusing it, but it works.

**Comment by eivindr** [comment]
Seems like the design elements is not properly signed. I updated to 2.1.0 hoping that it would solve the issue. But now I see that #45 was resolved after the 2.1.0 release.

Any chance for a new release with the fix for #45 ?

https://github.com/OpenNTF/org.openntf.nsfodp/commit/12b35120fbca8516f06de156dc555d2db4f466a9

**Comment by jesse-gallagher** [comment]
It’s in the works, but, between some feature creeps and unexpected snags, it’s been taking a long while. Maybe I’ll see if I can cobble together a milestone build.

**Comment by eivindr** [comment]
After upgrading to 3.0.0+bazar the design elements are now properly signed. Since we changed the server.id we do not need the option to specify the signer during the nsfodp stage. So for us this issue is resolved.

Thanks!
